Tx gasLimit > 100000000 causes gas limit reached error and stuck tx

## Tx gasLimit > 100000000 causes gas limit reached error and stuck tx

### Description

It seems that setting a tx gasLimit higher than 100000000 results in a pending tx that will never confirm. Any txs after this from the same account will also get stuck with no way of overwriting them.

https://github.com/0xPolygon/polygon-sdk/blob/335e78199b52c768e9b00c3f955e9a42a3ce5ed4/consensus/ibft/ibft.go#L356

Looking at the code above, a block header is initialized with its gasLimit set to 100000000. All tx gasLimits within that same block must not surpass this limit or else this issue occurs.

Aside from these transactions getting 'stuck', I observed many occurrences of the following log messages:
```
[ERROR] polygon: failed to apply tx: err=\"gas limit reached in the pool\""}

[31mERROR\u001b[0m\tbasichost\tfailed to resolve local interface addresses\t{\"error\": \"route ip+net: netlinkrib: too many open files\"}"}

[ERROR] polygon.network: failed to dial bootnode: err=\"peer rejected; insufficient capacity\""}
```

### Your environment

* OS and version `Ubuntu 20`
* version of the Polygon SDK `develop`
* branch that causes this issue `96be904`

### Steps to reproduce

* Tell us how to reproduce this issue <br />
* Where the issue is, if you know <br />
* Which commands triggered the issue, if any

1. Send a tx with gasLimit 5,000,000,000
2. Observe the `gas limit reached in the pool` log messages
3. Observe that subsequent txs from the same address cannot be processed
4. Attempting to override the nonce of the stuck tx (with a lower gasLimit but higher gasPrice) had no effect

### Expected behaviour

I expect there to be some sort of threshold that a tx's gasLimit must be under before it can be accepted. This threshold should be well under the hardcoded header gasLimit (aka Transition.gasPool) of 100000000.

### Actual behaviour

A tx with an extremely high gasLimit still gets picked up resulting in indefinite pending txs and error messages.

--- a/consensus/dev/dev.go
+++ b/consensus/dev/dev.go
@@ -138,18 +138,23 @@ func (d *Dev) writeNewBlock(parent *types.Header) error {
 		if txn == nil {
 			break
 		}
-		// Execute the state transition
-		if err := transition.Write(txn); err != nil {
-			if err.IsRecoverable {
-				retFn()
-			} else {
-				d.txpool.DecreaseAccountNonce(txn)
+
+		if txn.ExceedsBlockGasLimit(header.GasLimit) {
+			d.txpool.DecreaseAccountNonce(txn)
+		} else {
+			// Execute the state transition
+			if err := transition.Write(txn); err != nil {
+				if err.IsRecoverable {
+					retFn()
+				} else {
+					d.txpool.DecreaseAccountNonce(txn)
+				}
+
+				break
 			}
 
-			break
+			txns = append(txns, txn)
 		}
-
-		txns = append(txns, txn)
 	}
 
 	// Commit the changes

--- a/consensus/dev/dev.go
+++ b/consensus/dev/dev.go
@@ -140,7 +140,6 @@ func (d *Dev) writeNewBlock(parent *types.Header) error {
 		}
 		// Execute the state transition
 		if err := transition.Write(txn); err != nil {
-			// TODO : Update this for obvious reasons...
 			if err.IsRecoverable {
 				retFn()
 			} else {

--- a/consensus/dev/dev.go
+++ b/consensus/dev/dev.go
@@ -138,10 +138,14 @@ func (d *Dev) writeNewBlock(parent *types.Header) error {
 		if txn == nil {
 			break
 		}
-
 		// Execute the state transition
 		if err := transition.Write(txn); err != nil {
-			retFn()
+			// TODO : Update this for obvious reasons...
+			if err.IsRecoverable {
+				retFn()
+			} else {
+				d.txpool.DecreaseAccountNonce(txn)
+			}
 
 			break
 		}

--- a/consensus/ibft/ibft.go
+++ b/consensus/ibft/ibft.go
@@ -389,7 +389,11 @@ func (i *Ibft) buildBlock(snap *Snapshot, parent *types.Header) (*types.Block, e
 			break
 		}
 		if err := transition.Write(txn); err != nil {
-			retFn()
+			if err.IsRecoverable {
+				retFn()
+			} else {
+				i.txpool.DecreaseAccountNonce(txn)
+			}
 			break
 		}
 		txns = append(txns, txn)

--- a/consensus/ibft/ibft.go
+++ b/consensus/ibft/ibft.go
@@ -388,15 +388,20 @@ func (i *Ibft) buildBlock(snap *Snapshot, parent *types.Header) (*types.Block, e
 		if txn == nil {
 			break
 		}
-		if err := transition.Write(txn); err != nil {
-			if err.IsRecoverable {
-				retFn()
-			} else {
-				i.txpool.DecreaseAccountNonce(txn)
+
+		if txn.ExceedsBlockGasLimit(header.GasLimit) {
+			i.txpool.DecreaseAccountNonce(txn)
+		} else {
+			if err := transition.Write(txn); err != nil {
+				if err.IsRecoverable {
+					retFn()
+				} else {
+					i.txpool.DecreaseAccountNonce(txn)
+				}
+				break
 			}
-			break
+			txns = append(txns, txn)
 		}
-		txns = append(txns, txn)
 	}
 	i.logger.Info("picked out txns from pool", "num", len(txns), "remaining", i.txpool.Length())
 

--- a/e2e/txpool_test.go
+++ b/e2e/txpool_test.go
@@ -279,6 +279,13 @@ func TestTxPool_TransactionCoalescing(t *testing.T) {
 }
 
 func TestInvalidTransactionRecover(t *testing.T) {
+	// Test scenario :
+	//
+	// 1. Send a transaction with gasLimit > block gas limit.
+	//		-> The transaction should not be applied, and the nonce should not be incremented.
+	//
+	// 2. Send a second transaction with a gasLimit < block gas limit.
+	//		-> The transaction should be applied, and the receiver balance increased.
 	senderKey, senderAddress := tests.GenerateKeyAndAddr(t)
 	_, receiverAddress := tests.GenerateKeyAndAddr(t)
 	testTable := []struct {

--- a/e2e/txpool_test.go
+++ b/e2e/txpool_test.go
@@ -4,6 +4,7 @@ import (
 	"context"
 	"crypto/ecdsa"
 	"errors"
+	"github.com/umbracle/go-web3"
 	"io"
 	"math/big"
 	"testing"
@@ -275,4 +276,83 @@ func TestTxPool_TransactionCoalescing(t *testing.T) {
 		toAccountBalance.String(),
 		"To address balance mismatch after gap transaction",
 	)
+}
+
+func TestInvalidTransactionRecover(t *testing.T) {
+	senderKey, senderAddress := tests.GenerateKeyAndAddr(t)
+	_, receiverAddress := tests.GenerateKeyAndAddr(t)
+	testTable := []struct {
+		name              string
+		sender            types.Address
+		receiver          types.Address
+		nonce             uint64
+		submittedGasLimit uint64
+		value             *big.Int
+		shouldSucceed     bool
+	}{
+		{
+			name:              "Invalid transfer caused by exceeding block gas limit",
+			sender:            senderAddress,
+			receiver:          receiverAddress,
+			nonce:             0,
+			submittedGasLimit: 5000000000,
+			value:             oneEth,
+			shouldSucceed:     false,
+		},
+		{
+			name:              "Valid transfer #1",
+			sender:            senderAddress,
+			receiver:          receiverAddress,
+			nonce:             0,
+			submittedGasLimit: framework.DefaultGasLimit,
+			value:             oneEth,
+			shouldSucceed:     true,
+		},
+	}
+
+	servers := framework.NewTestServers(t, 1, func(config *framework.TestServerConfig) {
+		config.SetConsensus(framework.ConsensusDev)
+		config.SetSeal(true)
+		config.SetDevInterval(5)
+		config.Premine(senderAddress, framework.EthToWei(100))
+	})
+
+	server := servers[0]
+	client := server.JSONRPC()
+	operator := server.TxnPoolOperator()
+	ctx := context.Background()
+
+	for _, testCase := range testTable {
+		t.Run(testCase.name, func(t *testing.T) {
+			tx, err := signer.SignTx(&types.Transaction{
+				Nonce:    testCase.nonce,
+				GasPrice: big.NewInt(10000),
+				Gas:      testCase.submittedGasLimit,
+				To:       &testCase.receiver,
+				Value:    testCase.value,
+				V:        []byte{1},
+				From:     testCase.sender,
+			}, senderKey)
+			assert.NoError(t, err, "failed to sign transaction")
+
+			_, err = operator.AddTxn(ctx, &txpoolOp.AddTxnReq{
+				Raw: &any.Any{
+					Value: tx.MarshalRLP(),
+				},
+				From: types.ZeroAddress.String(),
+			})
+			assert.NoError(t, err, "failed to add txn using operator")
+
+			_ = waitForBlock(t, server, 1, 0)
+
+			balance, err := client.Eth().GetBalance(web3.Address(receiverAddress), web3.Latest)
+			assert.NoError(t, err, "failed to retrieve receiver account balance")
+
+			if testCase.shouldSucceed {
+				assert.Equal(t, oneEth.String(), balance.String())
+			} else {
+				assert.Equal(t, framework.EthToWei(0).String(), balance.String())
+			}
+		})
+	}
 }

--- a/e2e/txpool_test.go
+++ b/e2e/txpool_test.go
@@ -356,3 +356,106 @@ func TestInvalidTransactionRecover(t *testing.T) {
 		})
 	}
 }
+
+func TestTxPool_RecoverableError(t *testing.T) {
+	// Test scenario :
+	//
+	// 1. Send a first valid transaction with gasLimit = block gas limit - 1
+	//
+	// 2. Send a second transaction with gasLimit = block gas limit / 2. Since there is not enough gas remaining,
+	// the transaction will be pushed back to the pending queue so that is can be executed in the next block.
+	//
+	// 3. Send a third - valid - transaction, both the previous one and this one should be executed.
+	//
+	senderKey, senderAddress := tests.GenerateKeyAndAddr(t)
+	_, receiverAddress := tests.GenerateKeyAndAddr(t)
+	testTable := []struct {
+		name              string
+		sender            types.Address
+		senderKey         *ecdsa.PrivateKey
+		receiver          types.Address
+		nonce             uint64
+		gas               uint64
+		value             *big.Int
+		shouldSucceed     bool
+	}{
+		{
+			name:              "Valid transaction #1",
+			sender:            senderAddress,
+			senderKey:         senderKey,
+			receiver:          receiverAddress,
+			nonce:             0,
+			gas:               framework.DefaultGasLimit - 1,
+			value:             oneEth,
+			shouldSucceed:     true,
+		},
+		{
+			name:              "Invalid transaction (no enough gas remaining in pool)",
+			sender:            senderAddress,
+			senderKey:         senderKey,
+			receiver:          receiverAddress,
+			nonce:             1,
+			gas:               framework.DefaultGasLimit / 2,
+			value:             oneEth,
+			shouldSucceed:     false,
+		},
+		{
+			name:              "Valid transaction #2",
+			sender:            senderAddress,
+			senderKey:         senderKey,
+			receiver:          receiverAddress,
+			nonce:             2,
+			gas:               framework.DefaultGasLimit / 2,
+			value:             oneEth,
+			shouldSucceed:     true,
+		},
+	}
+
+	servers := framework.NewTestServers(t, 1, func(config *framework.TestServerConfig) {
+		config.SetConsensus(framework.ConsensusDev)
+		config.SetSeal(true)
+		config.SetDevInterval(5)
+		config.Premine(senderAddress, framework.EthToWei(100))
+	})
+
+	server := servers[0]
+	client := server.JSONRPC()
+	operator := server.TxnPoolOperator()
+	ctx := context.Background()
+
+	for _, testCase := range testTable {
+		t.Run(testCase.name, func(t *testing.T) {
+			tx, err := signer.SignTx(&types.Transaction{
+				Nonce:    testCase.nonce,
+				GasPrice: big.NewInt(10000),
+				Gas:      testCase.gas,
+				To:       &testCase.receiver,
+				Value:    testCase.value,
+				V:        []byte{1},
+				From:     testCase.sender,
+			}, testCase.senderKey)
+			assert.NoError(t, err, "failed to sign transaction")
+
+			_, err = operator.AddTxn(ctx, &txpoolOp.AddTxnReq{
+				Raw: &any.Any{
+					Value: tx.MarshalRLP(),
+				},
+				From: types.ZeroAddress.String(),
+			})
+			assert.NoError(t, err, "failed to add txn using operator")
+
+			if testCase.shouldSucceed {
+				_ = waitForBlock(t, server, 1, 0)
+			}
+		})
+	}
+
+	balance, err := client.Eth().GetBalance(web3.Address(receiverAddress), web3.Latest)
+	assert.NoError(t, err, "failed to retrieve receiver account balance")
+
+	assert.Equal(t, framework.EthToWei(3).String(), balance.String())
+
+	blockNumber, err := client.Eth().BlockNumber()
+	assert.NoError(t, err, "failed to retrieve most recent block number")
+	assert.Equal(t, blockNumber, uint64(2))
+}

--- a/state/executor.go
+++ b/state/executor.go
@@ -370,61 +370,35 @@ func (t *Transition) apply(msg *types.Transaction) (*runtime.ExecutionResult, *T
 
 	// 1. the nonce of the message caller is correct
 	if err := t.nonceCheck(msg); err != nil {
-		return nil, &TransitionApplicationError{
-			Err:           err,
-			IsRecoverable: true,
-		}
+		return nil, NewTransitionApplicationError(err, true)
 	}
 
 	// 2. caller has enough balance to cover transaction fee(gaslimit * gasprice)
 	if err := t.subGasLimitPrice(msg); err != nil {
-		return nil, &TransitionApplicationError{
-			Err:           err,
-			IsRecoverable: true,
-		}
-	}
-
-	// the amount of gas required is lower than the specified block gas limit
-	if msg.Gas > t.gasPool {
-		return nil, &TransitionApplicationError{
-			Err:           ErrBlockLimitExceeded,
-			IsRecoverable: false,
-		}
+		return nil, NewTransitionApplicationError(err, true)
 	}
 
 	// 3. the amount of gas required is available in the block
 	if err := t.subGasPool(msg.Gas); err != nil {
-		return nil, &TransitionApplicationError{
-			Err:           err,
-			IsRecoverable: true,
-		}
+		return nil, NewTransitionApplicationError(err, true)
 	}
 
 	// 4. there is no overflow when calculating intrinsic gas
 	intrinsicGasCost, err := TransactionGasCost(msg, t.config.Homestead, t.config.Istanbul)
 	if err != nil {
-		return nil, &TransitionApplicationError{
-			Err:           err,
-			IsRecoverable: true,
-		}
+		return nil, NewTransitionApplicationError(err, true)
 	}
 
 	// 5. the purchased gas is enough to cover intrinsic usage
 	gasLeft := msg.Gas - intrinsicGasCost
 	// Because we are working with unsigned integers for gas, the `>` operator is used instead of the more intuitive `<`
 	if gasLeft > msg.Gas {
-		return nil, &TransitionApplicationError{
-			Err:           ErrNotEnoughIntrinsicGas,
-			IsRecoverable: true,
-		}
+		return nil, NewTransitionApplicationError(err, true)
 	}
 
 	// 6. caller has enough balance to cover asset transfer for **topmost** call
 	if balance := txn.GetBalance(msg.From); balance.Cmp(msg.Value) < 0 {
-		return nil, &TransitionApplicationError{
-			Err:           runtime.ErrInsufficientBalance,
-			IsRecoverable: true,
-		}
+		return nil, NewTransitionApplicationError(err, true)
 	}
 
 	gasPrice := new(big.Int).Set(msg.GasPrice)

--- a/state/executor.go
+++ b/state/executor.go
@@ -184,7 +184,7 @@ func (t *Transition) Receipts() []*types.Receipt {
 var emptyFrom = types.Address{}
 
 // Write writes another transaction to the executor
-func (t *Transition) Write(txn *types.Transaction) *TransactionApplicationError {
+func (t *Transition) Write(txn *types.Transaction) *TransitionApplicationError {
 	signer := crypto.NewSigner(t.config, uint64(t.r.config.ChainID))
 
 	var err error
@@ -192,7 +192,7 @@ func (t *Transition) Write(txn *types.Transaction) *TransactionApplicationError 
 		// Decrypt the from address
 		txn.From, err = signer.Sender(txn)
 		if err != nil {
-			return &TransactionApplicationError{
+			return &TransitionApplicationError{
 				Err:           err,
 				IsRecoverable: false,
 			}
@@ -281,7 +281,7 @@ func (t *Transition) GetTxnHash() types.Hash {
 }
 
 // Apply applies a new transaction
-func (t *Transition) Apply(msg *types.Transaction) (result *runtime.ExecutionResult, err *TransactionApplicationError) {
+func (t *Transition) Apply(msg *types.Transaction) (result *runtime.ExecutionResult, err *TransitionApplicationError) {
 	s := t.state.Snapshot()
 	result, err = t.apply(msg)
 	if err != nil {
@@ -339,16 +339,23 @@ var (
 	ErrNotEnoughFunds        = fmt.Errorf("not enough funds for transfer with given value")
 )
 
-type TransactionApplicationError struct {
+type TransitionApplicationError struct {
 	Err				error
 	IsRecoverable	bool // Should the transaction be discarded, or put back in the queue.
 }
 
-func (e *TransactionApplicationError) Error() string {
+func (e *TransitionApplicationError) Error() string {
 	return fmt.Sprintf("%v, recoverable [%t]", e.Err, e.IsRecoverable)
 }
 
-func (t *Transition) apply(msg *types.Transaction) (*runtime.ExecutionResult, *TransactionApplicationError) {
+func NewTransitionApplicationError(err error, isRecoverable bool) *TransitionApplicationError {
+	return &TransitionApplicationError{
+		Err:           err,
+		IsRecoverable: isRecoverable,
+	}
+}
+
+func (t *Transition) apply(msg *types.Transaction) (*runtime.ExecutionResult, *TransitionApplicationError) {
 	// First check this message satisfies all consensus rules before
 	// applying the message. The rules include these clauses
 	//
@@ -363,7 +370,7 @@ func (t *Transition) apply(msg *types.Transaction) (*runtime.ExecutionResult, *T
 
 	// 1. the nonce of the message caller is correct
 	if err := t.nonceCheck(msg); err != nil {
-		return nil, &TransactionApplicationError{
+		return nil, &TransitionApplicationError{
 			Err:           err,
 			IsRecoverable: true,
 		}
@@ -371,7 +378,7 @@ func (t *Transition) apply(msg *types.Transaction) (*runtime.ExecutionResult, *T
 
 	// 2. caller has enough balance to cover transaction fee(gaslimit * gasprice)
 	if err := t.subGasLimitPrice(msg); err != nil {
-		return nil, &TransactionApplicationError{
+		return nil, &TransitionApplicationError{
 			Err:           err,
 			IsRecoverable: true,
 		}
@@ -379,7 +386,7 @@ func (t *Transition) apply(msg *types.Transaction) (*runtime.ExecutionResult, *T
 
 	// the amount of gas required is lower than the specified block gas limit
 	if msg.Gas > t.gasPool {
-		return nil, &TransactionApplicationError{
+		return nil, &TransitionApplicationError{
 			Err:           ErrBlockLimitExceeded,
 			IsRecoverable: false,
 		}
@@ -387,7 +394,7 @@ func (t *Transition) apply(msg *types.Transaction) (*runtime.ExecutionResult, *T
 
 	// 3. the amount of gas required is available in the block
 	if err := t.subGasPool(msg.Gas); err != nil {
-		return nil, &TransactionApplicationError{
+		return nil, &TransitionApplicationError{
 			Err:           err,
 			IsRecoverable: true,
 		}
@@ -396,7 +403,7 @@ func (t *Transition) apply(msg *types.Transaction) (*runtime.ExecutionResult, *T
 	// 4. there is no overflow when calculating intrinsic gas
 	intrinsicGasCost, err := TransactionGasCost(msg, t.config.Homestead, t.config.Istanbul)
 	if err != nil {
-		return nil, &TransactionApplicationError{
+		return nil, &TransitionApplicationError{
 			Err:           err,
 			IsRecoverable: true,
 		}
@@ -406,7 +413,7 @@ func (t *Transition) apply(msg *types.Transaction) (*runtime.ExecutionResult, *T
 	gasLeft := msg.Gas - intrinsicGasCost
 	// Because we are working with unsigned integers for gas, the `>` operator is used instead of the more intuitive `<`
 	if gasLeft > msg.Gas {
-		return nil, &TransactionApplicationError{
+		return nil, &TransitionApplicationError{
 			Err:           ErrNotEnoughIntrinsicGas,
 			IsRecoverable: true,
 		}
@@ -414,7 +421,7 @@ func (t *Transition) apply(msg *types.Transaction) (*runtime.ExecutionResult, *T
 
 	// 6. caller has enough balance to cover asset transfer for **topmost** call
 	if balance := txn.GetBalance(msg.From); balance.Cmp(msg.Value) < 0 {
-		return nil, &TransactionApplicationError{
+		return nil, &TransitionApplicationError{
 			Err:           runtime.ErrInsufficientBalance,
 			IsRecoverable: true,
 		}

--- a/state/executor.go
+++ b/state/executor.go
@@ -184,7 +184,7 @@ func (t *Transition) Receipts() []*types.Receipt {
 var emptyFrom = types.Address{}
 
 // Write writes another transaction to the executor
-func (t *Transition) Write(txn *types.Transaction) error {
+func (t *Transition) Write(txn *types.Transaction) *TransactionApplicationError {
 	signer := crypto.NewSigner(t.config, uint64(t.r.config.ChainID))
 
 	var err error
@@ -192,17 +192,20 @@ func (t *Transition) Write(txn *types.Transaction) error {
 		// Decrypt the from address
 		txn.From, err = signer.Sender(txn)
 		if err != nil {
-			return err
+			return &TransactionApplicationError{
+				Err:           err,
+				IsRecoverable: false,
+			}
 		}
 	}
 
 	// Make a local copy and apply the transaction
 	msg := txn.Copy()
 
-	result, err := t.Apply(msg)
-	if err != nil {
-		t.logger.Error("failed to apply tx", "err", err)
-		return err
+	result, e := t.Apply(msg)
+	if e != nil {
+		t.logger.Error("failed to apply tx", "err", e)
+		return e
 	}
 	t.totalGas += result.GasUsed
 
@@ -278,7 +281,7 @@ func (t *Transition) GetTxnHash() types.Hash {
 }
 
 // Apply applies a new transaction
-func (t *Transition) Apply(msg *types.Transaction) (result *runtime.ExecutionResult, err error) {
+func (t *Transition) Apply(msg *types.Transaction) (result *runtime.ExecutionResult, err *TransactionApplicationError) {
 	s := t.state.Snapshot()
 	result, err = t.apply(msg)
 	if err != nil {
@@ -330,12 +333,22 @@ var (
 	ErrNonceIncorrect        = fmt.Errorf("incorrect nonce")
 	ErrNotEnoughFundsForGas  = fmt.Errorf("not enough funds to cover gas costs")
 	ErrBlockLimitReached     = fmt.Errorf("gas limit reached in the pool")
+	ErrBlockLimitExceeded     = fmt.Errorf("specified gas exceed block gas limit")
 	ErrIntrinsicGasOverflow  = fmt.Errorf("overflow in intrinsic gas calculation")
 	ErrNotEnoughIntrinsicGas = fmt.Errorf("not enough gas supplied for intrinsic gas costs")
 	ErrNotEnoughFunds        = fmt.Errorf("not enough funds for transfer with given value")
 )
 
-func (t *Transition) apply(msg *types.Transaction) (result *runtime.ExecutionResult, err error) {
+type TransactionApplicationError struct {
+	Err				error
+	IsRecoverable	bool // Should the transaction be discarded, or put back in the queue.
+}
+
+func (e *TransactionApplicationError) Error() string {
+	return fmt.Sprintf("%v, recoverable [%t]", e.Err, e.IsRecoverable)
+}
+
+func (t *Transition) apply(msg *types.Transaction) (*runtime.ExecutionResult, *TransactionApplicationError) {
 	// First check this message satisfies all consensus rules before
 	// applying the message. The rules include these clauses
 	//
@@ -349,36 +362,62 @@ func (t *Transition) apply(msg *types.Transaction) (result *runtime.ExecutionRes
 	txn := t.state
 
 	// 1. the nonce of the message caller is correct
-	if err = t.nonceCheck(msg); err != nil {
-		return
+	if err := t.nonceCheck(msg); err != nil {
+		return nil, &TransactionApplicationError{
+			Err:           err,
+			IsRecoverable: true,
+		}
 	}
 
 	// 2. caller has enough balance to cover transaction fee(gaslimit * gasprice)
-	if err = t.subGasLimitPrice(msg); err != nil {
-		return
+	if err := t.subGasLimitPrice(msg); err != nil {
+		return nil, &TransactionApplicationError{
+			Err:           err,
+			IsRecoverable: true,
+		}
+	}
+
+	// the amount of gas required is lower than the specified block gas limit
+	if msg.Gas > t.gasPool {
+		return nil, &TransactionApplicationError{
+			Err:           ErrBlockLimitExceeded,
+			IsRecoverable: false,
+		}
 	}
 
 	// 3. the amount of gas required is available in the block
-	if err = t.subGasPool(msg.Gas); err != nil {
-		return
+	if err := t.subGasPool(msg.Gas); err != nil {
+		return nil, &TransactionApplicationError{
+			Err:           err,
+			IsRecoverable: true,
+		}
 	}
 
 	// 4. there is no overflow when calculating intrinsic gas
 	intrinsicGasCost, err := TransactionGasCost(msg, t.config.Homestead, t.config.Istanbul)
 	if err != nil {
-		return
+		return nil, &TransactionApplicationError{
+			Err:           err,
+			IsRecoverable: true,
+		}
 	}
 
 	// 5. the purchased gas is enough to cover intrinsic usage
 	gasLeft := msg.Gas - intrinsicGasCost
 	// Because we are working with unsigned integers for gas, the `>` operator is used instead of the more intuitive `<`
 	if gasLeft > msg.Gas {
-		return nil, ErrNotEnoughIntrinsicGas
+		return nil, &TransactionApplicationError{
+			Err:           ErrNotEnoughIntrinsicGas,
+			IsRecoverable: true,
+		}
 	}
 
 	// 6. caller has enough balance to cover asset transfer for **topmost** call
 	if balance := txn.GetBalance(msg.From); balance.Cmp(msg.Value) < 0 {
-		return nil, runtime.ErrInsufficientBalance
+		return nil, &TransactionApplicationError{
+			Err:           runtime.ErrInsufficientBalance,
+			IsRecoverable: true,
+		}
 	}
 
 	gasPrice := new(big.Int).Set(msg.GasPrice)
@@ -388,6 +427,7 @@ func (t *Transition) apply(msg *types.Transaction) (result *runtime.ExecutionRes
 	t.ctx.GasPrice = types.BytesToHash(gasPrice.Bytes())
 	t.ctx.Origin = msg.From
 
+	var result *runtime.ExecutionResult = nil
 	if msg.IsContractCreation() {
 		result = t.Create2(msg.From, msg.Input, value, gasLeft)
 	} else {

--- a/state/executor.go
+++ b/state/executor.go
@@ -192,10 +192,7 @@ func (t *Transition) Write(txn *types.Transaction) *TransitionApplicationError {
 		// Decrypt the from address
 		txn.From, err = signer.Sender(txn)
 		if err != nil {
-			return &TransitionApplicationError{
-				Err:           err,
-				IsRecoverable: false,
-			}
+			return NewTransitionApplicationError(err, false)
 		}
 	}
 
@@ -333,7 +330,7 @@ var (
 	ErrNonceIncorrect        = fmt.Errorf("incorrect nonce")
 	ErrNotEnoughFundsForGas  = fmt.Errorf("not enough funds to cover gas costs")
 	ErrBlockLimitReached     = fmt.Errorf("gas limit reached in the pool")
-	ErrBlockLimitExceeded     = fmt.Errorf("specified gas exceed block gas limit")
+	ErrBlockLimitExceeded     = fmt.Errorf("specified gas exceeds block gas limit")
 	ErrIntrinsicGasOverflow  = fmt.Errorf("overflow in intrinsic gas calculation")
 	ErrNotEnoughIntrinsicGas = fmt.Errorf("not enough gas supplied for intrinsic gas costs")
 	ErrNotEnoughFunds        = fmt.Errorf("not enough funds for transfer with given value")

--- a/state/executor.go
+++ b/state/executor.go
@@ -330,15 +330,15 @@ var (
 	ErrNonceIncorrect        = fmt.Errorf("incorrect nonce")
 	ErrNotEnoughFundsForGas  = fmt.Errorf("not enough funds to cover gas costs")
 	ErrBlockLimitReached     = fmt.Errorf("gas limit reached in the pool")
-	ErrBlockLimitExceeded     = fmt.Errorf("specified gas exceeds block gas limit")
+	ErrBlockLimitExceeded    = fmt.Errorf("specified gas exceeds block gas limit")
 	ErrIntrinsicGasOverflow  = fmt.Errorf("overflow in intrinsic gas calculation")
 	ErrNotEnoughIntrinsicGas = fmt.Errorf("not enough gas supplied for intrinsic gas costs")
 	ErrNotEnoughFunds        = fmt.Errorf("not enough funds for transfer with given value")
 )
 
 type TransitionApplicationError struct {
-	Err				error
-	IsRecoverable	bool // Should the transaction be discarded, or put back in the queue.
+	Err           error
+	IsRecoverable bool // Should the transaction be discarded, or put back in the queue.
 }
 
 func (e *TransitionApplicationError) Error() string {

--- a/txpool/txpool.go
+++ b/txpool/txpool.go
@@ -225,8 +225,10 @@ func (t *TxPool) addImpl(ctx string, tx *types.Transaction) error {
 // Since any discarded transaction should not affect the world state, the nextNonce should be reset to the value
 // it was set to before the transaction appeared.
 func (t *TxPool) DecreaseAccountNonce(tx *types.Transaction) {
-	txnsQueue := t.accountQueues[tx.From]
-	txnsQueue.nextNonce -= 1
+	if t.accountQueues[tx.From] != nil {
+		txnsQueue := t.accountQueues[tx.From]
+		txnsQueue.nextNonce -= 1
+	}
 }
 
 // GetTxs gets both pending and queued transactions

--- a/txpool/txpool.go
+++ b/txpool/txpool.go
@@ -219,6 +219,16 @@ func (t *TxPool) addImpl(ctx string, tx *types.Transaction) error {
 	return nil
 }
 
+// DecreaseAccountNonce resets the nonce attached to an account whenever a transaction produce an error which is not
+// recoverable, meaning the transaction will be discarded.
+//
+// Since any discarded transaction should not affect the world state, the nextNonce should be reset to the value
+// it was set to before the transaction appeared.
+func (t *TxPool) DecreaseAccountNonce(tx *types.Transaction) {
+	txnsQueue, _ := t.accountQueues[tx.From]
+	txnsQueue.nextNonce -= 1
+}
+
 // GetTxs gets both pending and queued transactions
 func (t *TxPool) GetTxs() (map[types.Address]map[uint64]*types.Transaction, map[types.Address]map[uint64]*types.Transaction) {
 

--- a/txpool/txpool.go
+++ b/txpool/txpool.go
@@ -225,7 +225,7 @@ func (t *TxPool) addImpl(ctx string, tx *types.Transaction) error {
 // Since any discarded transaction should not affect the world state, the nextNonce should be reset to the value
 // it was set to before the transaction appeared.
 func (t *TxPool) DecreaseAccountNonce(tx *types.Transaction) {
-	txnsQueue, _ := t.accountQueues[tx.From]
+	txnsQueue := t.accountQueues[tx.From]
 	txnsQueue.nextNonce -= 1
 }
 

--- a/types/transaction.go
+++ b/types/transaction.go
@@ -63,3 +63,7 @@ func (t *Transaction) Cost() *big.Int {
 	total.Add(total, t.Value)
 	return total
 }
+
+func (t *Transaction) ExceedsBlockGasLimit(blockGasLimit uint64) bool {
+	return t.Gas > blockGasLimit
+}

